GVRf/Framework/framework/src/main/jni/objects/lightlist.cpp

```diff
@@ -0,0 +1,372 @@
+    if (it3 != mClassMap.end())
+    {
+        std::vector<Light*>& lights = it3->second;
+        auto it2 = std::find(lights.begin(), lights.end(), light);
+        if (it2 == lights.end())
+        {
+            return false;
+        }
+        light->setLightIndex(-1);
+        /*
+         * If all lights in the class are gone,
+         * remove the class from the map.
+         */
+        if (lights.size() == 0)
+        {
+            mClassMap.erase(it3);
+            return true;
+        }
+        /*
+         * Removed a light, recompute light indices for all
+         * lights of that type
+         */
+        else
+        {
+            int index = 0;
+            lights.erase(it2);
+            for (auto it = lights.begin();
+                 it != lights.end();
+                 ++it)
+            {
+                Light* l = *it;
+                l->setLightIndex(index++);
+            }
+        }
+    }
+#ifdef DEBUG_LIGHT
+    LOGD("LIGHT: %s removed from scene", light->getLightClass());
+#endif
+    mDirty |= LIGHT_REMOVED | REBUILD_SHADERS;
+    return true;
+}
+
+void LightList::shadersRebuilt()
+{
+    mDirty &= ~REBUILD_SHADERS;
+}
+
+void LightList::forEachLight(std::function<void(Light&)> func)
+{
+    for (auto it1 = mClassMap.begin();
+         it1 != mClassMap.end();
+         ++it1)
+    {
+        const std::vector<Light*>& lights = it1->second;
+        for (auto it2 = lights.begin(); it2 != lights.end(); ++it2)
+        {
+            Light* light = *it2;
+            if (light)
+            {
+                func(*light);
+            }
+        }
+    }
+}
+
+void LightList::forEachLight(std::function<void(const Light&)> func) const
+{
+    for (auto it1 = mClassMap.begin();
+         it1 != mClassMap.end();
+         ++it1)
+    {
+        const std::vector<Light*>& lights = it1->second;
+        for (auto it2 = lights.begin(); it2 != lights.end(); ++it2)
+        {
+            Light* light = *it2;
+            if (light)
+            {
+                func(*light);
+            }
+        }
+    }
+}
+
+ShadowMap* LightList::scanLights()
+{
+    ShadowMap* shadowMap = NULL;
+
+    mTotalUniforms = 0;
+    for (auto it1 = mClassMap.begin();
+         it1 != mClassMap.end();
+         ++it1)
+    {
+        const std::vector<Light*>& lights = it1->second;
+        for (auto it2 = lights.begin(); it2 != lights.end(); ++it2)
+        {
+            Light* light = *it2;
+
+            mTotalUniforms += light->uniforms().uniforms().getNumEntries();
+            ShadowMap* sm = light->getShadowMap();
+            if (sm && sm->enabled())
+            {
+                shadowMap = sm;
+            }
+        }
+    }
+    return shadowMap;
+}
+
+ShadowMap* LightList::updateLightBlock(Renderer* renderer)
+{
+    std::lock_guard < std::recursive_mutex > lock(mLock);
+    bool dirty = mDirty != 0;
+    ShadowMap* shadowMap = NULL;
+
+    if (mDirty & REBUILD_SHADERS)
+    {
+        return scanLights();
```

Comment:
This function should not return from here. There are cases when we add a light, both the flags REBUILD_SHADERS and LIGHT_ADDED are set.